woocommerce/google-listings-and-ads
# Product feed API endpoint

### Changes proposed in this Pull Request:

Part of #481 .

This PR adds the API endpoint for retrieving the store's product feed.
- Adds the `GET /wp-json/wc/gla/mc/product-feed` endpoint, with a result of:
    ```json
    {
	    "products": [
	        {
	            "id": 24,
	            "title": "Album",
	            "visible": true,
	            "status": "has-errors",
	            "errors": [
	                "[imageLink] This value should not be blank."
	            ]
	        },
	        {
	            "id": 193,
	            "title": "Beanie",
	            "visible": true,
	            "status": "synced",
	            "errors": []
	        },
	        "…"
        ],
        "total": 23,
        "page": "3"
    }
    ```
- `products` is the array of store products with some details for the table:
    - `id` and `title` are product details. `id` can be used to generate the `Edit` link for the product.
    - `visible` is a bool that maps to the Don't/Sync and Show option of the product.
    - `status` is a new meta item for the product (see below) and can be `pending`, `synced`, `not-synced`, or `has-errors`.
    - `errors` is an array of errors found for the product _prior_ to attempting a sync with the Merchant Center. Variable products will show all errors of their variations.
- `total` and `page` for pagination.
- The endpoint also accepts several query parameters:
  - `per_page` and `page` - used to paginate the results.
  - `query` - used to filter by product title.
  - `orderby` and `order` - used to sort the results:
    - `orderby` accepts the values `title` (default), `id`, `visible`, and `status`.
    - `order` accepts the values `asc` and `desc`.
- The endpoint only returns `simple` and `variable` products, but _NOT_ `variations`. So instead of `Sweater - Large`, `Sweater - Medium` and `Sweater - Small`, there's just a `Sweater` item.
  - Variable products without any variations will NOT be displayed.
- The PR adds a `postmeta` item for each product: `_wc_gla_sync_status`, which has the same values as the `status` attribute in the response.
    - The new meta value is updated realtime with the product and reflects the extension's knowledge of the product status. A later PR may add the Merchant Center-reported status as well.
- This PR also modifies the Product update process to fill empty `visibility` values when a product is marked as `synced` or `invalid`, in order to facilitate ordering by `visibility`
- Creates the `DB/ProductFeedQueryHelper` class to handle the query preparation for the product feed
  - ~Specifically, it uses the `posts_where` and `posts_orderby` hooks to enable title filtering and result status ordering.~
  - It also receives `wpdb` as a dependency, which is why it's in the `DB` namespace.

<details>
<summary><em>See full REST response</em></summary>

```json
# Order by status desc
{
    "products": [
        {
            "id": 18,
            "title": "Cap",
            "visible": true,
            "status": "synced",
            "errors": []
        },
        {
            "id": 23,
            "title": "Polo",
            "visible": true,
            "status": "pending",
            "errors": []
        },
        {
            "id": 22,
            "title": "Long Sleeve Tee",
            "visible": false,
            "status": "not-synced",
            "errors": []
        },
        {
            "id": 24,
            "title": "Album",
            "visible": true,
            "status": "has-errors",
            "errors": [
                "[imageLink] This value should not be blank.",   
                "[price] This value should not be null."
            ]
        }
    ],
    "total": 4,
    "page": 1
}
```

</details>

### Detailed test instructions:

1. Create a product or two before activating GLA (so they shouldn't have an GLA-related meta options).
2. Set up a Merchant Center account (using the onboarding wizard).
2. Create some products and sync them (they're queued to sync when created), or sync existing products using the button on _Connection Test_.
3. Send a request to `GET /wp-json/wc/gla/mc/product-feed`
  - Confirm results look correct
  - Test the different query params and confirm they work as expected
  - Modify some products (set to don't sync, remove required values, etc) and see that the changes are reflected in the feed (after giving them time to be synced/checked).
